pkg/prometheus/agent/operator.go

```diff
@@ -564,6 +585,86 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 
 	p := pobj.(*monitoringv1alpha1.PrometheusAgent)
 	p = p.DeepCopy()
+	if p.Spec.Mode != nil && *p.Spec.Mode == "DaemonSet" {
+		err = c.syncDaemonSet(ctx, key, p)
+	} else {
+		err = c.syncStatefulSet(ctx, key, p)
+	}
+	c.reconciliations.SetStatus(key, err)
+	return err
+}
+
+func (c *Operator) syncDaemonSet(ctx context.Context, key string, p *monitoringv1alpha1.PrometheusAgent) error {
+	if !c.daemonSetFeatureGateEnabled {
```

Comment:
After more thinking I wonder if we shouldn't fallback to "statefulset" sync when the feature gate isn't enabled instead of failing?

Reply:
Hm, you mean if I set `mode: daemonSet`, but forgot to enable feature gate, I'll have statefulset? :thinking:

Reply:
Yes. WDYT?

Reply:
I not a fan of the idea 😬

If someone explicitly tells the operator to run a daemonset, see no errors but see statefulset instead... that would be very confusing

Reply:
We've discussed offline and agreed that failing would be better (more explicit).